feat(ivy): throw compilation error when providing undecorated classes

crisbeto opened the pull request:
Adds a compilation error if the consumer tries to pass in an undecorated class into the `providers` of an `NgModule`, or the `providers`/`viewProviders` arrays of a `Directive`/`Component`.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -170,12 +174,18 @@ export class ComponentDecoratorHandler implements
 
-    const viewProviders: Expression|null = component.has('viewProviders') ?
+    const rawViewProviders = component.get('viewProviders');
+    const wrappedViewProviders: Expression|null = rawViewProviders ?
```
In general @alxhub does not like us to rely upon the result of a `get()` call to determine whether an item is in a map, preferring to explicitly use `has()`.

petebacondarwin replied:
Did you mean to 👍 this comment and then not change the code ? 😺

crisbeto replied:
I kept it here, because `rawViewProviders` was used in 4 other places within the same block. I've changed it everywhere else though.

alxhub replied:
Sorry, but @petebacondarwin is right. We should explicitly use `.has` even for `rawViewProviders`:

```typescript
const rawViewProviders = component.has('viewProviders') ? component.get('viewProviders') ! : null;
```

In general I think this code could benefit from fewer ternaries. Maybe structured like:

```typescript
let providersRequiringFactory: Set<ClassDeclaration>|null = null;

if (component.has('providers')) {
  // populate providersRequiringFactory if needed
}

let viewProviders: Expression[]|null = null;
if (component.has('viewProviders')) {
  // append view providers that require factories to providersRequiringFactory
}
```

crisbeto replied:
Done. Also added a comment about why `providersRequiringFactory` and `viewProvidersRequiringFactory` are kept separately.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -489,7 +504,28 @@ export class ComponentDecoratorHandler implements
+      result.diagnostics = result.diagnostics || [];
+      const providerDiagnostics = getProviderDiagnostics(
+          analysis.providersRequiringFactory, analysis.meta.providers !.node, this.factoryRegistry);
+      if (providerDiagnostics) {
```
The result of `getProviderDiagnostics()` is always an array, so I expect this to always be true.
Similarly below.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -98,16 +105,35 @@ export class DirectiveDecoratorHandler implements
+      if (providerDiagnostics) {
+        diagnostics = providerDiagnostics;
+      }
+    }
```
This block is pretty much duplicated three times: once here and twice in the component `resolve()` method.
Could it be moved into a helper, which would also help to decrease the LOC in the already very long component `resolve()` method.

crisbeto replied:
I noticed it as I was writing it as well, but it was also fairly trivial code so pulling it out into a function probably won't help much. I wrote it this way, because I noticed other places where we return `undefined` instead of an empty array of diagnostics. If it's fine to return an empty array, we can simplify it a lot.

petebacondarwin replied:
I think it would help in having less brain-load when reading the `resolve()` methods :-)

I think it would be great if we could avoid the `undefined` when dealing with diagnostics. There doesn't really appear to be any real difference between `undefined` and `[]` in the code. It is probably just that the `diagnostics` property on `ResolveResult` is optional.

petebacondarwin replied:
Similarly did you intend to make changes to this?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,44 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+}
+
+/**
+ * Resolves the classes, passed into a Provider[], that are supposed to have a factory definition.
```
```suggestion
 * Resolves the given `rawProviders` into `ClassDeclarations` and returns a set containing those that are known to require a factory definition.
```

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,44 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+      // Skip declarations from external libraries since the consumer can't do much about them.
+      if (!node.getSourceFile().isDeclarationFile && isNamedClassDeclaration(node)) {
+        const constructor = node.members.find(ts.isConstructorDeclaration);
+        if (constructor !== undefined && constructor.parameters.length > 0) {
```
This is the key line in this function IMO.
A comment here pointing out that we are capturing this `node` because it has a contructor and that constructor has arguments, which will need injecting, would be helpful for future readers.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/metadata/src/registry.ts:
```diff
@@ -59,3 +59,14 @@ export class CompoundMetadataRegistry implements MetadataRegistry {
 }
+
+/**
+ * Keeps track of class declarations that have factories
```
This description (and the name of the registry) threw me for a while (and maybe I am still confused).
As I understand it, this is a registry of "injectable things" (e.g. services, directives, components, etc) that we know about, right? The fact that these injectables may require a factory to be able to instantiate them in DI is not actually relevant here.

crisbeto replied:
Good point. I had a hard time figuring out how to name the class. I'll think about it a bit more.

petebacondarwin commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -5222,6 +5222,244 @@ export const Foo = Foo__PRE_R3__;
     });
 
+    describe('undecorated providers', () => {
+      it('should error when an undecorated class is provided directly in a module', () => {
```
I feel that we need to be more clear here, since a class that has no constructor or a constructor with no parameters IS allowed.

```suggestion
      it('should error when an undecorated class, with a non-trivial constructor, is provided directly in a module', () => {
```

petebacondarwin reviewed (changes requested):
In the tests there is no mention of undecorated classes that have no constructor but inherit from a base class that does have a non-trivial constructor. Is this case covered by this PR?

There are a few minor changes suggested. Also the PR is conflicted and needs a rebase.

crisbeto commented:
@petebacondarwin the case when a class inherits a non-trivial constructor from an undecorated base class is handled by https://github.com/angular/angular/pull/34061.

crisbeto committed:
feat(ivy): throw compilation error when providing undecorated classes

Adds a compilation error if the consumer tries to pass in an undecorated class into the `providers` of an `NgModule`, or the `providers`/`viewProviders` arrays of a `Directive`/`Component`.

crisbeto commented:
Rebased and addressed the feedback from @petebacondarwin.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,48 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+        const constructor = node.members.find(ts.isConstructorDeclaration);
+
+        // Note that we only want to capture providers with a non-trivial constructor,
+        // because they're the ones that might be using DI and need to be decorated.
```
👍

alxhub commented on packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import {ConstantPool} from '@angular/compiler';
+import {InjectableClassRegistry} from '@angular/compiler-cli/src/ngtsc/metadata/src/registry';
```
Relative import path.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -170,11 +175,17 @@ export class ComponentDecoratorHandler implements
+            this.annotateForClosureCompiler ? wrapFunctionExpressionsInParens(rawViewProviders) :
+                                              rawViewProviders) :
+        null;
+    const viewProvidersRequiringFactory = rawViewProviders ?
```
Use explicit comparisons: `rawViewProviders !== null`.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/registry.ts:
```diff
@@ -59,3 +59,13 @@ export class CompoundMetadataRegistry implements MetadataRegistry {
+export class InjectableClassRegistry {
+  private classes = new Set<ClassDeclaration>();
+
+  registerClass(declaration: ClassDeclaration): void { this.classes.add(declaration); }
```
Let's go with `registerInjectable` and `isInjectable` to make the code which consumes `InjectableClassRegistry` even more obvious about what it's doing.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,48 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+      }
+    };
+
+    resolvedProviders.forEach(function processProviders(provider) {
```
Use a `for...of` loop instead.

I'm trying to switch the compiler over to these as they make debugging a lot simpler.

crisbeto replied:
Using a `for...of` here would require more code to do the recursion below.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,48 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+        }
+      } else if (Array.isArray(provider)) {
+        // If we ran into an array, recurse into it until we've resolve all the classes.
+        provider.forEach(processProviders);
```
Oooh. This is clever.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,48 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+      if (provider instanceof Reference) {
+        handleReference(provider);
+      } else if (provider instanceof Map) {
+        const useExisting = provider.get('useClass');
```
Check `.has` first.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,48 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+      const node = reference.node;
+
+      // Skip declarations from external libraries since the consumer can't do much about them.
+      if (!node.getSourceFile().isDeclarationFile && isNamedClassDeclaration(node)) {
```
I think we should definitely handle the case someone does `providers: [Service]` and `Service` comes from an external library.

The `InjectableClassRegistry` can look if it has a `ɵfac` field present, and consider it an injectable if so.

crisbeto replied:
I think that removing this check will end up introducing breaking changes somewhere. I added it in the first place, because it was erroring for [this example](https://github.com/angular/components/blob/master/src/components-examples/cdk/scrolling/cdk-virtual-scroll-custom-strategy/cdk-virtual-scroll-custom-strategy-example.ts#L16) in the CDK which in turn inherits [from here](https://github.com/angular/components/blob/master/src/cdk/scrolling/fixed-size-virtual-scroll.ts#L41) that has a non-trivial constructor with `number` parameters. We can fix the CDK quickly since it's something we control, but I can imagine other libraries out there that have the same problem.

crisbeto committed:
fixup! feat(ivy): throw compilation error when providing undecorated classes

crisbeto commented:
Addressed/replied to all of the feedback.

crisbeto committed:
fixup! feat(ivy): throw compilation error when providing undecorated classes

crisbeto committed:
fixup! feat(ivy): throw compilation error when providing undecorated classes

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:
```diff
@@ -0,0 +1,36 @@
+    registry: InjectableClassRegistry): ts.Diagnostic[] {
+  const diagnostics: ts.Diagnostic[] = [];
+
+  providerClasses.forEach(provider => {
```
Prefer a `for...of` loop (it's easier to debug through).

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:
```diff
@@ -0,0 +1,36 @@
+
+  providerClasses.forEach(provider => {
+    if (!registry.isInjectable(provider)) {
+      diagnostics.push(makeDiagnostic(
```
Let's do a trick here :)

1) Capture the `Reference` instead of the `ClassDeclaration` when tracking these providers.
2) Pass through the `ts.SourceFile` for the component/directive (let's call it `hostFile`).
3) Call `Reference.getIdentityIn(hostFile)`. If it returns a `ts.Identifier`, check whether the `ts.Identifier` is within the `ts.Expression` for the providers (which you call `highlightNode`). You can do this check positionally, by checking the `pos`/`end` of each.
4) If the result of `getIdentityIn` is contained in the `highlightNode`, then it's the identifier by which the class was referenced in the providers array, and we can put the diagnostic there. Otherwise we can fall back on the `providers` array itself.

crisbeto replied:
Cool! I don't think we need to pass in the `SourceFile`, because we can take it off of the `highlightNode` (I'll rename it to something more appropriate).

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,46 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+      if (provider instanceof Reference) {
+        handleReference(provider);
+      } else if (provider instanceof Map && provider.has('useClass')) {
+        const useExisting = provider.get('useClass');
```
```suggestion
        const useExisting = provider.get('useClass') !;
```

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,46 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+  if (Array.isArray(resolvedProviders)) {
+    const handleReference = (reference: Reference) => {
+      const node = reference.node;
+      if (isNamedClassDeclaration(node)) {
```
Maybe use `ReflectionHost.getConstructorParameters` instead of the TS APIs directly.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,46 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+
+    resolvedProviders.forEach(function processProviders(provider) {
+      if (provider instanceof Reference) {
+        handleReference(provider);
```
I think it would be cleaner if instead of a callback, we first pick out the `reference` depending on the shape of the provider, and then process it inline.

```typescript
let tokenClass: Reference<ClassDeclaration>|null = null;

if (provider instanceof Reference) {
  tokenClass = provider;
} else if (provider instanceof Map && provider.has('useClass')) {
  const useClass = provider.get('useClass') !;
  if (useClass instanceof Reference) {
    tokenClass = useClass;
  }
}

if (tokenClass === null) {
  return;
}

// process the tokenClass here.
```

crisbeto committed:
fixup! feat(ivy): throw compilation error when providing undecorated classes

crisbeto commented:
Addressed the last set of feedback.

crisbeto commented:
Addressed the latest set of feedback.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:
```diff
@@ -0,0 +1,50 @@
+  const diagnostics: ts.Diagnostic[] = [];
+
+  for (const provider of providerClasses) {
+    if (!registry.isInjectable(provider.node)) {
```
Nit: rather than have a deeper `if`, we can invert the condition:

```typescript
for (const provider of providerClasses) {
  if (registry.isInjectable(provider.node)) {
    continue;
  }

  // Now logic here proceeds as normal, without the need for nesting.
}

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,46 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+  const providers = new Set<Reference<ClassDeclaration>>();
+  const resolvedProviders = evaluator.evaluate(rawProviders);
+
+  if (Array.isArray(resolvedProviders)) {
```
Same here - we could avoid nesting with an early return here, rather than nesting the entire function body in an `if` block.

alxhub commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:285804135:BASE:285804235:1576522177734:ca39b808)
[Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:285804135:BASE:285804235:1576522177703:6356269f)

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -375,4 +375,46 @@ const parensWrapperTransformerFactory: ts.TransformerFactory<ts.Expression> =
+        provider.forEach(processProviders);
+      } else if (provider instanceof Reference) {
+        tokenClass = provider;
+      } else if (provider instanceof Map && provider.has('useClass')) {
```
The g3 presubmit brought up an interesting case:

```typescript
{provide: Foo, useClass: Bar, deps: [DepA, DepB]}
```

is perfectly fine even if `Bar` is not injectable, as the user is explicitly telling the injector how to call the constructor. So `useClass` with `deps` can be ignored here:

```suggestion
      } else if (provider instanceof Map && provider.has('useClass') && !provider.has('deps')) {
```

Please add a test to that effect :)

crisbeto committed:
fixup! feat(ivy): throw compilation error when providing undecorated classes

crisbeto commented:
Fixed the presubmit issue, addressed the feedback and rebased.

alxhub commented:
Updated:

[Presubmit](http://test/OCL:285804135:BASE:286080470:1576627432631:4ebb8553)
[Ivy Presubmit](http://test/OCL:285804135:BASE:286080470:1576627432631:3cbb15ef)

alxhub commented:
Closing in favor of #34460.
